agent/consul/gateway_locator.go

```diff
@@ -0,0 +1,284 @@
+package consul
+
+import (
+	"math/rand"
+	"sort"
+	"sync"
+
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/ipaddr"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/go-hclog"
+	memdb "github.com/hashicorp/go-memdb"
+)
+
+// GatewayLocator assists in selecting an appropriate mesh gateway when wan
+// federation via mesh gateways is enabled.
+//
+// NOTE: this is ONLY for RPC and gossip so we only need to track local/primary
+type GatewayLocator struct {
+	logger            hclog.Logger
+	srv               serverDelegate
+	datacenter        string // THIS dc
+	primaryDatacenter string
+
+	// these ONLY contain ones that have the wanfed:1 meta
+	gatewaysLock    sync.Mutex
+	primaryGateways []string // WAN addrs
+	localGateways   []string // LAN addrs
+
+	// primaryMeshGatewayDiscoveredAddresses is the current fallback addresses
+	// for the mesh gateways in the primary datacenter.
+	primaryMeshGatewayDiscoveredAddresses     []string
+	primaryMeshGatewayDiscoveredAddressesLock sync.Mutex
+
+	// This will be closed the FIRST time we get some gateways populated
+	primaryGatewaysReadyCh chan struct{}
+}
+
+// PrimaryMeshGatewayAddressesReadyCh returns a channel that will be closed
+// when federation state replication ships back at least one primary mesh
+// gateway (not via fallback config).
+func (g *GatewayLocator) PrimaryMeshGatewayAddressesReadyCh() <-chan struct{} {
+	return g.primaryGatewaysReadyCh
+}
+
+// PickGateway returns the address for a gateway suitable for reaching the
+// provided datacenter.
+func (g *GatewayLocator) PickGateway(dc string) string {
+	item := g.pickGateway(dc == g.primaryDatacenter)
+	g.logger.Trace("picking gateway for transit", "gateway", item, "source_datacenter", g.datacenter, "dest_datacenter", dc)
+	return item
+}
+
+func (g *GatewayLocator) pickGateway(primary bool) string {
+	addrs := g.listGateways(primary)
+	return getRandomItem(addrs)
+}
+
+func (g *GatewayLocator) listGateways(primary bool) []string {
+	g.gatewaysLock.Lock()
+	defer g.gatewaysLock.Unlock()
+
+	var addrs []string
+	if primary {
+		addrs = g.primaryGateways
+	} else {
+		addrs = g.localGateways
+	}
+
+	if primary && len(addrs) == 0 {
+		addrs = g.PrimaryGatewayFallbackAddresses()
+	}
+
+	return addrs
+}
+
+// RefreshPrimaryGatewayFallbackAddresses is used to update the list of current
+// fallback addresses for locating mesh gateways in the primary datacenter.
```

Comment:
Maybe worth noting the semantics of the int returned. It's not super obvious to me why the caller would care if the update was effectively a no-op or not or why they'd need to know the current number of addrs if it was  changed considering that they passed that array in the first place. I.e. why isn't it just a bool?

Reply:
It's got the weird signature so that it can work in a retryJoin.join callback. I'm going to clarify this by confining the weirdness just to where that callback is used instead of spreading it all over.